File: Java/src/main/java/Solution101_200/Solution160.java
package Solution101_200;

import tool.ListNode;

import java.util.HashMap;
import java.util.HashSet;

/**
 * @author windows
 */
public class Solution160 {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        HashSet<ListNode> set = new HashSet<>();
        while(headA != null) {
            set.add(headA);
            headA = headA.next;
        }
        while(headB != null) {
            if (set.contains(headB)) {
                return headB;
            }
            headB = headB.next;
        }
        return null;
    }


    public static void main(String[] args) {
        Solution160 solution160 = new Solution160();
    }
}

File: C++/src/Solution101_200/Solution160.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution160.h"
#include <cstring>

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    if (headA == nullptr || headB == nullptr){
      return nullptr;
    }
    ListNode* a = headA, *b = headB;
    while (a != b) {
      a == nullptr ? headB : a->next;
      b == nullptr ? headA : b->next;
    }
    return a;
}
